gobblin-data-management/src/main/java/gobblin/data/management/retention/dataset/HiveCleanableDataset.java

```diff
@@ -0,0 +1,115 @@
+import gobblin.data.management.version.finder.VersionFinder;
+import gobblin.hive.HiveMetastoreClientPool;
+import gobblin.util.AutoReturnableObject;
+import gobblin.util.Either;
+
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * Extends {@link HiveDataset} and also is cleanable. It uses {@link CombineRetentionPolicy} of {@link HiveTablePartitionTimeBasedRetentionPolicy}
+ * and {@link HiveTablePartitionInvalidDataLocRetentionPolicy} to decide the deletable versions.
+ */
+@Slf4j
+public class HiveCleanableDataset extends HiveDataset implements CleanableDataset {
+  private final VersionFinder<HiveTablePartitionVersion> versionFinder;
+  private final CombineRetentionPolicy<HiveTablePartitionVersion> retentionPolicy;
+  private final boolean simulate;
+
+  public HiveCleanableDataset(FileSystem fs, HiveMetastoreClientPool clientPool, Table table, Properties properties)
+      throws IOException {
+    super(fs, clientPool, table, properties);
+    this.versionFinder = new HiveDatasetVersionFinder(fs, properties);
+    List<RetentionPolicy<HiveTablePartitionVersion>> policies = Lists
+        .newArrayList(new HiveTablePartitionInvalidDataLocRetentionPolicy(fs),
+            new HiveTablePartitionTimeBasedRetentionPolicy(properties, fs));
+    this.retentionPolicy = new CombineRetentionPolicy<HiveTablePartitionVersion>(policies,
+        CombineRetentionPolicy.DeletableCombineOperation.UNION);
+    this.simulate = Boolean.valueOf(properties
+        .getProperty(MultiVersionCleanableDatasetBase.SIMULATE_KEY, MultiVersionCleanableDatasetBase.SIMULATE_DEFAULT));
+  }
+
+  @Override
+  public void clean()
+      throws IOException {
+    Iterable<Either<Table, Partition>> deletableTableOrPartitions = Iterables.transform(
+        this.retentionPolicy.listDeletableVersions(new ArrayList<>(this.versionFinder.findDatasetVersions(this))),
+        new Function<HiveTablePartitionVersion, Either<Table, Partition>>() {
+          @Override
+          public Either<Table, Partition> apply(HiveTablePartitionVersion input) {
+            return input.getVersion();
+          }
+        });
+    for (Either<Table, Partition> tableOrPartitionToDelete : deletableTableOrPartitions) {
+      this.cleanImpl(tableOrPartitionToDelete);
+    }
+  }
+
+  protected void cleanImpl(Either<Table, Partition> tableOrPartition)
+      throws IOException {
+    try (AutoReturnableObject<IMetaStoreClient> client = this.clientPool.getClient()) {
+      if (tableOrPartition instanceof Either.Right) {
+        Partition paritionToDrop = ((Partition) ((Either.Right) tableOrPartition).getRight());
+        if (this.simulate) {
+          log.info("Simulate mode: will delete partition " + paritionToDrop.getValues() + " for table " + paritionToDrop
+              .getTable().getTableName());
+        }
+
+        client.get().dropPartition(this.table.getDbName(), this.table.getTableName(), paritionToDrop.getValues(), true);
```

Comment:
Should this be in an `else` clause? If I'm not reading wrong, right now simulate will log the message and then do the deletion either way.